src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509Certificate2.cs

```diff
@@ -778,6 +779,116 @@ public X509Certificate2 CopyWithPrivateKey(ECDiffieHellman privateKey)
+
+            if (algorithm is null)
+            {
+                return null;
+            }
+
+            byte[] publicKey = Pal.PublicKeyValue;
+
+            if (publicKey.Length != algorithm.PublicKeySizeInBytes)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            return MLDsa.ImportMLDsaPublicKey(algorithm, Pal.PublicKeyValue);
+        }
+
+        /// <summary>
+        ///   Get the <see cref="MLDsa"/> private key from this certificate.
+        /// </summary>
+        /// <returns>
+        ///   The private key, or <see langword="null"/> if this certificate does not have an ML-DSA private key.
+        /// </returns>
+        /// <exception cref="CryptographicException">
+        ///   An error occurred accessing the private key.
+        /// </exception>
+        [Experimental(Experimentals.PostQuantumCryptographyDiagId)]
+        public MLDsa? GetMLDsaPrivateKey()
+        {
+            MLDsaAlgorithm? algorithm = MLDsaAlgorithm.GetMLDsaAlgorithmFromOid(GetKeyAlgorithm());
+
+            if (algorithm is null)
+            {
+                return null;
+            }
+
+            return Pal.GetMLDsaPrivateKey();
+        }
+
+        /// <summary>
+        ///   Combines a private key with a certificate containing the associated public key into a
+        ///   new instance that can access the private key.
+        /// </summary>
+        /// <param name="privateKey">
+        ///   The ML-DSA private key that corresponds to the ML-DSA public key in this certificate.
+        /// </param>
+        /// <returns>
+        ///   A new certificate with the <see cref="HasPrivateKey" /> property set to <see langword="true"/>.
+        ///   The current certificate isn't modified.
+        /// </returns>
+        /// <exception cref="ArgumentNullException">
+        ///   <paramref name="privateKey"/> is <see langword="null"/>.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        ///   The specified private key doesn't match the public key for this certificate.
+        /// </exception>
+        /// <exception cref="InvalidOperationException">
+        ///   The certificate already has an associated private key.
+        /// </exception>
+        [Experimental(Experimentals.PostQuantumCryptographyDiagId)]
+        public X509Certificate2 CopyWithPrivateKey(MLDsa privateKey)
+        {
+            ArgumentNullException.ThrowIfNull(privateKey);
+
+            if (HasPrivateKey)
+                throw new InvalidOperationException(SR.Cryptography_Cert_AlreadyHasPrivateKey);
+
+            using (MLDsa? publicKey = GetMLDsaPublicKey())
+            {
+                if (publicKey is null)
+                {
+                    throw new ArgumentException(SR.Cryptography_PrivateKey_WrongAlgorithm);
+                }
+
+                if (publicKey.Algorithm != privateKey.Algorithm)
```

Comment:
Do we _really_ want a reference equality check here? I understand that the algorithm is a set of closed values from properties, but it always makes me a little nervous.

Alternatively, add `IEquatable` and the comparison operations on `MLDsaAlgorithm` like `MLKemAlgorithm` has.

Reply:
When ML-KEM lands for API Review we'll have the IEquatable vs not discussion, and then all of the algorithm identifier types should treat them equally.

My personal take is that reference equality is correct.... at least for Core.  For Framework I guess there's the possibility that someone wants to use this across a remoting boundary, and so it's a clone instead of a consolidated reference.  I just sort of shrugged it off to API Review when you proposed it before :)

Reply:
I have strong opinions, but, sure can wait until API review.